### DISCUSSION TITLE
How to make next-auth User Session Data Globally accessible from _app.js server-side?

### DISCUSSION THREAD

**Discussion by yelnyafacee** [nextauthjs/next-auth#2069] [Help]
What is the best way to make `next-auth` `user` data in `session` globally accessible in all pages server-side? 

or do I need to manually add `getServerSideProps` to every page to get the `session` `user` data?

`<Provider session={pageProps.session}></Provider>` is in `_app.js` `return` function so do I need a `hoc` `wrapper` component for this?

I tried using `getInitialProps`:

```

function MyApp({ Component, pageProps, initialLoginStatus }) {

   console.log(initialLoginStatus);

   ....

  return (

     <React.Fragment>

          <Provider session={pageProps.session}>
             ...
          </Provider>

     </React.Fragment>

  )
}

MyApp.getInitialProps = async ( ctx ) => {

  if (typeof window === "undefined") {

    return {
      initialLoginStatus: `test`
    }

  } else {

    return {
      initialLoginStatus: `error`
    }
  }

}
```
but I get this error: `TypeError: Cannot read property 'session' of undefined`
for `<Provider session={pageProps.session}>`

**Comment by balazsorban44** [comment]
Could you explain a bit better what do you mean by "globally accessible server-side"?

The [Provider](https://next-auth.js.org/getting-started/client#provider) component makes sure that subsequent requests through `useSession()` will be served from this "global" [React Context](https://reactjs.org/docs/context.html), if the session is not deemed stale yet. Otherwise it will visit the backend, and ask for a new session.

Once you added `Provider` to your `_app`, you can use `useSession()` at any level, on any page.

Keep in mind, that pages that use `getServerSideProps` are always rendered on-demand, so you might need to return `session` in their `props` object, so `_app` can receive it through `pageProps.session`

**Reply by yelnyafacee** [reply]
I want to display the username + profile picture in the top menu bar if the user is logged in and hide them if they are not, instead of adding useSession() in getServerSideProps on every page, is there a way to add it to _app.js only once server-side?

**Reply by balazsorban44** [reply]
`useSession()` in `getServerSideProps` is not a thing, since useSession can only be used in a React Component.

Since you don't know which route will be the initial entrypoint to your app by the user, you must return the session in each page's `getServerSideProps`

**Reply by yelnyafacee** [reply]
I see, so using `getServerSideProps` for all pages is the only way?

**Reply by balazsorban44** [reply]
We probably misunderstand each other. If you simply want to show a logged-in header, it does not mean you have to add `getServerSideProps` to ALL your pages. Have a look at our example page to get a better idea: https://github.com/nextauthjs/next-auth-example

**Reply by yelnyafacee** [reply]
Hi, my issue with `const [ session, loading ] = useSession()` is if I `refresh` the page with browser, I can see the top `logged-in header` flickers between `login/register` and `you are logged in as : xxx`.

is it possible to prevent that?

**Reply by balazsorban44** [reply]
Since it loads the session from client-side, no. In that case, you will actually have to use getServerSideProps nevertheless. I would reevaluate though, if a "flicker" should make the whole site blocking, or you could use some kind of skeleton UI until the session is loaded? That way the flicker will be much gentler and the page won't jump. but that's just my opinion. if you absolutely want to avoid a state where the user is not shown for second and you are aware of the performance consequences, go ahead and use getServerSideProps, on all pages that you expect the user to start the page from.

**Comment by vonsamnang01041994** [comment]
On Wed, 26 May 2021 at 19:53 yelnya ***@***.***> wrote:

> I see, so using getServerSideProps for all pages is the only way?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nextauthjs/next-auth/discussions/2069#discussioncomment-786142>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOWDUMTVNK6HNHDNCLTS3NTTPTVLPANCNFSM45RNXJCA>
> .
>
-- 
Agreement
